src/Microsoft.Diagnostics.Runtime/src/Builders/SegmentBuilder.cs

```diff
@@ -10,31 +10,47 @@ namespace Microsoft.Diagnostics.Runtime.Builders
     internal sealed class SegmentBuilder : ISegmentData
     {
         private SegmentData _segment;
+        private bool _regions;
         private ulong _heapAllocated;
-        private ulong _ephemGen0Start;
-        private ulong _ephemGen1Start;
         private ulong _ephemAddress;
         private readonly SOSDac _sos;
 
+        // Segments only
+        private ulong _ephemGen0Start;
+        private ulong _ephemGen1Start;
+
+        // Regions only
+        private int _generation;
+
         public SegmentBuilder(SOSDac sos)
         {
             _sos = sos;
         }
 
-        public bool Initialize(ulong address, in HeapDetails heap)
+        public bool Initialize(ulong address, int generation, in HeapDetails heap)
         {
+            _regions = heap.SavedSweepEphemeralSeg.Value == -1;
+
             _heapAllocated = heap.Allocated;
-            _ephemGen0Start = heap.GenerationTable[0].AllocationStart;
-            _ephemGen1Start = heap.GenerationTable[1].AllocationStart;
+            if (_regions)
+            {
+                _generation = generation;
+            }
+            else
+            {
+                _ephemGen0Start = heap.GenerationTable[0].AllocationStart;
+                _ephemGen1Start = heap.GenerationTable[1].AllocationStart;
+            }
             _ephemAddress = heap.EphemeralHeapSegment;
 
             return _sos.GetSegmentData(address, out _segment);
         }
 
         #region ISegmentData
         public int LogicalHeap { get; set; }
-
-        public ulong BaseAddress => _segment.Address;
+        
+        // 0x20 is sizeof(aligned_plug_and_gap)
+        public ulong BaseAddress => _regions ? (_segment.Start - 0x20) : _segment.Address;
```

Comment:
Is `aligned_plug_and_gap` the same size on all architectures and platforms?

Reply:
It isn't, but as of now, we only support regions on 64-bit platforms, so we were fine. To futureproof the code, I just made an update to make the size calculation depends on the platform pointer size so that it will be automatically correct when we decide to support regions for x86 or other 32-bit platforms.